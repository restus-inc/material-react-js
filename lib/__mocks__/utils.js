/* global jest */
const utils = jest.genMockFromModule('../utils');

const newId = () => 'a0123456';
utils.newId = newId;

export default utils;
export { newId };
