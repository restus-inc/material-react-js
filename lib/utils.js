/**
 * Utilities for Material React.
 * @module material-react/lib/utils
 */

/**
 * Generates a rundom ID. The generated ID starts with basic latin child alphabet, and
 * subsequent characters consist of child alphanumeric character. (i.e. it matches
 * `/[a-z][a-z0-9]{digit-1}/`)
 * @param {number} [digit=8] The number of digit of the generated ID
 * @returns {string} The ID generated
 */
export function newId(digit = 8) {
  return (
    (Math.floor(Math.random() * 26) + 10) * (36 ** (digit - 1))
    + Math.floor(Math.random() * (36 ** (digit - 1)))
  ).toString(36);
}

/**
 * Converts the event name to event handler name.
 * @param {string} eventName The event name that is converted to the handler name
 * @return {string} The event handler name converted
 */
export function handlerName(eventName) {
  return `on${eventName.charAt(0).toUpperCase()}${eventName.slice(1)}`;
}
