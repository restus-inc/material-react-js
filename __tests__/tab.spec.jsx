/* global jest, describe, afterEach, it, expect */
import 'regenerator-runtime/runtime';
import React from 'react';
import { render, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { htmlOfRendering } from './utils';

import Tab from '../lib/tab';
import TabBar from '../lib/tab-bar';

describe('Tab and TabBar component', () => {
  afterEach(async () => {
    await cleanup();
  });

  it('render tab control', () => {
    expect(htmlOfRendering((
      <TabBar>
        <Tab label="foo" active={true}/>
        <Tab label="bar"/>
        <Tab label="baz"/>
      </TabBar>
    ))).resolves.toMatchSnapshot();

    expect(htmlOfRendering((
      <TabBar className="tab-bar">
        <Tab label="foo" className="tab" icon="grade" iconClassName="icon" />
        <Tab label="bar" icon="favorite"/>
        <Tab label="baz" className="tab" active={true}/>
      </TabBar>
    ))).resolves.toMatchSnapshot();
  });

  it('fires onActivated event when the tab is activated', () => {
    let eventDetail = null;
    const onActivated = jest.fn((event) => {
      eventDetail = event.detail;
    });

    const { getAllByRole } = render((
      <TabBar onActivated={onActivated}>
        <Tab label="foo" active={true}/>
        <Tab label="bar"/>
        <Tab label="baz"/>
      </TabBar>
    ));
    userEvent.click(getAllByRole('tab')[0]);
    expect(onActivated).not.toHaveBeenCalled();
    userEvent.click(getAllByRole('tab')[2]);
    expect(onActivated).toHaveBeenCalledTimes(1);
    expect(eventDetail).toEqual({ index: 2 });
  });
});
