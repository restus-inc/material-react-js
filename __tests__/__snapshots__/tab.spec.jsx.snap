// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab and TabBar component render tab control 1`] = `
"<div
  class=\\"mdc-tab-bar\\"
  role=\\"tablist\\"
>
  <div
    class=\\"mdc-tab-scroller\\"
  >
    <div
      class=\\"mdc-tab-scroller__scroll-area mdc-tab-scroller__scroll-area--scroll\\"
      style=\\"margin-bottom: 0px;\\"
    >
      <div
        class=\\"mdc-tab-scroller__scroll-content\\"
      >
        <button
          aria-selected=\\"true\\"
          class=\\"mdc-tab mdc-tab--active\\"
          id=\\"mdc-tab-1\\"
          role=\\"tab\\"
          tabindex=\\"-1\\"
          type=\\"button\\"
        >
          <span
            class=\\"mdc-tab__content\\"
          >
            <span
              class=\\"mdc-tab__text-label\\"
            >
              foo
            </span>
          </span>
          <span
            class=\\"mdc-tab-indicator mdc-tab-indicator--active\\"
          >
            <span
              class=\\"mdc-tab-indicator__content mdc-tab-indicator__content--underline\\"
            />
          </span>
          <span
            class=\\"mdc-tab__ripple\\"
          />
        </button>
        <button
          class=\\"mdc-tab\\"
          id=\\"mdc-tab-2\\"
          role=\\"tab\\"
          tabindex=\\"-1\\"
          type=\\"button\\"
        >
          <span
            class=\\"mdc-tab__content\\"
          >
            <span
              class=\\"mdc-tab__text-label\\"
            >
              bar
            </span>
          </span>
          <span
            class=\\"mdc-tab-indicator\\"
          >
            <span
              class=\\"mdc-tab-indicator__content mdc-tab-indicator__content--underline\\"
            />
          </span>
          <span
            class=\\"mdc-tab__ripple\\"
          />
        </button>
        <button
          class=\\"mdc-tab\\"
          id=\\"mdc-tab-3\\"
          role=\\"tab\\"
          tabindex=\\"-1\\"
          type=\\"button\\"
        >
          <span
            class=\\"mdc-tab__content\\"
          >
            <span
              class=\\"mdc-tab__text-label\\"
            >
              baz
            </span>
          </span>
          <span
            class=\\"mdc-tab-indicator\\"
          >
            <span
              class=\\"mdc-tab-indicator__content mdc-tab-indicator__content--underline\\"
            />
          </span>
          <span
            class=\\"mdc-tab__ripple\\"
          />
        </button>
      </div>
    </div>
  </div>
</div>"
`;

exports[`Tab and TabBar component render tab control 2`] = `
"<div
  class=\\"mdc-tab-bar tab-bar\\"
  role=\\"tablist\\"
>
  <div
    class=\\"mdc-tab-scroller\\"
  >
    <div
      class=\\"mdc-tab-scroller__scroll-area mdc-tab-scroller__scroll-area--scroll\\"
      style=\\"margin-bottom: 0px;\\"
    >
      <div
        class=\\"mdc-tab-scroller__scroll-content\\"
      >
        <button
          class=\\"mdc-tab tab\\"
          id=\\"mdc-tab-4\\"
          role=\\"tab\\"
          tabindex=\\"-1\\"
          type=\\"button\\"
        >
          <span
            class=\\"mdc-tab__content\\"
          >
            <span
              aria-hidden=\\"true\\"
              class=\\"icon mdc-tab__icon\\"
            >
              grade
            </span>
            <span
              class=\\"mdc-tab__text-label\\"
            >
              foo
            </span>
          </span>
          <span
            class=\\"mdc-tab-indicator\\"
          >
            <span
              class=\\"mdc-tab-indicator__content mdc-tab-indicator__content--underline\\"
            />
          </span>
          <span
            class=\\"mdc-tab__ripple\\"
          />
        </button>
        <button
          class=\\"mdc-tab\\"
          id=\\"mdc-tab-5\\"
          role=\\"tab\\"
          tabindex=\\"-1\\"
          type=\\"button\\"
        >
          <span
            class=\\"mdc-tab__content\\"
          >
            <span
              aria-hidden=\\"true\\"
              class=\\"material-icons mdc-tab__icon\\"
            >
              favorite
            </span>
            <span
              class=\\"mdc-tab__text-label\\"
            >
              bar
            </span>
          </span>
          <span
            class=\\"mdc-tab-indicator\\"
          >
            <span
              class=\\"mdc-tab-indicator__content mdc-tab-indicator__content--underline\\"
            />
          </span>
          <span
            class=\\"mdc-tab__ripple\\"
          />
        </button>
        <button
          aria-selected=\\"true\\"
          class=\\"mdc-tab mdc-tab--active tab\\"
          id=\\"mdc-tab-6\\"
          role=\\"tab\\"
          tabindex=\\"-1\\"
          type=\\"button\\"
        >
          <span
            class=\\"mdc-tab__content\\"
          >
            <span
              class=\\"mdc-tab__text-label\\"
            >
              baz
            </span>
          </span>
          <span
            class=\\"mdc-tab-indicator mdc-tab-indicator--active\\"
          >
            <span
              class=\\"mdc-tab-indicator__content mdc-tab-indicator__content--underline\\"
            />
          </span>
          <span
            class=\\"mdc-tab__ripple\\"
          />
        </button>
      </div>
    </div>
  </div>
</div>"
`;
